smartsim/_core/entrypoints/telemetrymonitor.py

```diff
@@ -55,17 +60,520 @@
 from smartsim._core.launcher.pbs.pbsLauncher import PBSLauncher
 from smartsim._core.launcher.slurm.slurmLauncher import SlurmLauncher
 from smartsim._core.launcher.stepInfo import StepInfo
-from smartsim._core.utils.helpers import get_ts
+from smartsim._core.utils.helpers import get_ts_ms
 from smartsim._core.utils.serialize import MANIFEST_FILENAME
 from smartsim.error.errors import SmartSimError
+from smartsim.log import get_logger
 from smartsim.status import STATUS_COMPLETED, TERMINAL_STATUSES
 
+# pylint: disable=too-many-lines
 """Telemetry Monitor entrypoint"""
 
 # kill is not catchable
 SIGNALS = [signal.SIGINT, signal.SIGQUIT, signal.SIGTERM, signal.SIGABRT]
 _EventClass = t.Literal["start", "stop", "timestep"]
 _MAX_MANIFEST_LOAD_ATTEMPTS: t.Final[int] = 6
+F_MIN, F_MAX = 1.0, 600.0
+_LOG_FILE_NAME = "logs/telemetrymonitor.out"
+
+logger = get_logger(__name__)
+
+
+class Sink(abc.ABC):
+    """Base class for telemetry output sinks"""
+
+    @abc.abstractmethod
+    async def save(self, *args: t.Any) -> None:
+        """Save the passed data to the underlying sink"""
+
+
+class FileSink(Sink):
+    """Telemetry sink that writes to a file"""
+
+    def __init__(self, filename: str) -> None:
+        """Initialize the FileSink
+        :param entity: The JobEntity producing log data
+        :type entity: JobEntity
+        :param filename: The relative path and filename of the file to be written
+        :type filename: str"""
+        super().__init__()
+        filename = self._check_init(filename)
+        self._path = pathlib.Path(filename)
+
+    @staticmethod
+    def _check_init(filename: str) -> str:
+        """Validate initialization arguments.
+        raise ValueError if an invalid filename is passed"""
+
+        if not filename:
+            raise ValueError("No filename provided to FileSink")
+
+        return filename
+
+    @property
+    def path(self) -> pathlib.Path:
+        """Returns the path to the underlying file the FileSink will write to"""
+        return self._path
+
+    async def save(self, *args: t.Any) -> None:
+        """Save all arguments to a file as specified by the associated JobEntity"""
+        self._path.parent.mkdir(parents=True, exist_ok=True)
+
+        async with await open_file(self._path, "a+", encoding="utf-8") as sink_fp:
+            values = ",".join(map(str, args)) + "\n"
+            await sink_fp.write(values)
+
+
+class Collector(abc.ABC):
+    """Base class for metrics collectors"""
+
+    def __init__(self, entity: JobEntity, sink: Sink) -> None:
+        """Initialize the collector
+
+        :param entity: The entity to collect metrics on
+        :type entity: JobEntity"""
+        self._entity = entity
+        self._sink = sink
+        self._enabled = True
```

Comment:
I can potentially see the need in the future for a list of sinks to log the data to multiple endpoints. Could we refactor this now before we have to untangle it in the future?